### ISSUE TITLE
Multiselect with complex objects

### ISSUE THREAD

**Issue by Luk164** [Megabit/Blazorise#2050]
**Is your feature request related to a problem? Please describe.**
https://github.com/stsrki/Blazorise/issues/1714

**Describe the solution you'd like**
A support for complex objects in multi-select combo boxes. One way to do this would be to create GUIDs for different objects and then handle the lists internally in blazorise as previously outlined in the issue above.

**Comment by niem94** [comment]
@stsrki Is there any current workaround? I want to use a MultiSelect in my DataGrid, but I am not sure how I can persist the selections as a list of selected objects when I use the approach in #1714.

Right now I have this:

```csharp
<Select TValue="string" Multiple="true" SelectedValues="@(((List<Role>)context.CellValue).Select(x => x.Id).ToList().AsReadOnly())" SelectedValuesChanged="@OnSelectedValuesChanged" SelectedValuesExpression="@(() => context.Item.Roles.Select(x => x.Id).ToList().AsReadOnly())">
    <ChildContent>
        @foreach (var role in EcosystemState.Value.CurrentEcosystem.Roles)
                {
            <SelectItem TValue="string" Value="role.Id" title="@role.Description">@role.Name </SelectItem>
        }
    </ChildContent>
    <Feedback>
        <ValidationError/>
    </Feedback>
</Select>

private void OnSelectedValuesChanged(IReadOnlyList<string> values)
{
    SelectedValues = values;            
}
```

Is the only option to have the list of Ids be available in the SavedRowItem command, or am I missing something?

**Comment by stsrki** [comment]
Yes, the only option is to have an ID for a Select/SelectItem value. Complex types are problematic because they cannot be serialized to the DOM element value. Only basic types.

**Comment by niem94** [comment]
@stsrki I do not see a way to make this work with validation, as I am not able to set the FieldIdentifier correctly. It requires the same object type as the SelectedValues list, but as the SelectedValues list only works with primitive types, I am not able to set the FieldIdentifier to point to the matching object and thus have no information on the actual property name and model.

Is it possible to set the SelectedValues to a List<string> that holds e.g. the ids of the elements in a list, while setting the SelectedValuesExpression to something that populated the FieldIdentifier with e.g. Name = "Id" and Model = MyObject?
If not do you have a workaround in mind, so I can have validation on Multi-Select until this issue is resolved.

**Comment by stsrki** [comment]
@niem94 I played a little and here is one example with complex type.

```cshtml
<Validations Model="@selectedModel">
    <Validation>
        <Select TValue="Customer" Multiple @bind-SelectedValues="@selectedModel.Customers">
            @foreach ( var customer in selectedModel.Customers)
            {
                <SelectItem @key="@customer" TValue="Customer" Value="@customer">@customer.Name</SelectItem>
            }
        </Select>
    </Validation>
</Validations>
@code{
    public class Model
    {
        public int ID { get; set; }
        public string Name { get; set; }
        public IReadOnlyList<Customer> Customers { get; set; } = new List<Customer>
        {
            new Customer { ID = 1, Name = "Jane" },
            new Customer { ID = 2, Name = "Rose" },
            new Customer { ID = 3, Name = "Paul" },
        };
    }

    public class Customer
    {
        public int ID { get; set; }
        public string Name { get; set; }

        public override string ToString()
        {
            return ID.ToString();
        }
    }

    Model selectedModel = new Model();
}
```

The most important part is the `public override string ToString()`. That way the object can be serialized into DOM element value. There are still some issue that I need to figure when selecting the items. But it compiles and renders at least 😅

**Comment by stsrki** [comment]
Found the problem. It is converted the other way. From DOM value which is number into the complex object. Unfortunately for now there is no way it seems. I will research more soon after I finish other work and see what can be done.

**Comment by niem94** [comment]
@stsrki Awesome ill try this 👍🏻 I think the ToString solution will solve my problem.
